src/custom_composer/callbacks.py

```diff
@@ -503,14 +499,96 @@ def after_backward(self, state: State, logger: Logger) -> None:
         # Log total gradient norm across all non-embedding parameters
         total_norm = total_norm ** (1.0 / 2)
         if self.include_embedding_params:
-            metrics["grad_norm/total"] = total_norm
+            metrics["grad_stats/norm"] = total_norm
         else:
-            metrics["grad_norm/total_non_embedding"] = total_norm
+            metrics["grad_stats/norm_non_embedding"] = total_norm
 
         if metrics:
             logger.log_metrics(metrics)
 
 
+class LogGradientVariance(Callback):
+    """Log variance over multiple gradient updates."""
+
+    def __init__(
+        self,
+        accumulation_steps: int = 10,
+        log_frequency: int = 1,
+        include_embedding_params: bool = False,
+        *args: Any,
+        **kwargs: Any,
+    ):
+        super().__init__(*args, **kwargs)
+        self.accumulation_steps = accumulation_steps
+        self.log_frequency = log_frequency
+        self.include_embedding_params = include_embedding_params
+        self.step_count = 0
+        self.accumulated_grads: list[dict[str, torch.Tensor]] = []
+
+    @staticmethod
+    def _is_embedding_param(param_name: str) -> bool:
+        """Check if a parameter is an embedding parameter."""
+        return "embed" in param_name.lower()
+
+    @staticmethod
+    def _get_model(state: State):
+        """Get the model, handling wrapped models."""
+        if hasattr(state.model, "module"):
+            return state.model.module.model
+        return state.model.model
+
+    def after_train_batch(self, state: State, logger: Logger) -> None:
+        """Accumulate gradients and log variance across steps."""
+        self.step_count += 1
+        if self.step_count % self.log_frequency != 0:
```

Comment:
Should we maybe use the step counter from `state` instead?

Also can you clarify what `log_frequency` refers to? My current understanding is that every `log_frequency` steps you add the accumulated gradients. Would it maybe be cleaner to scrap `log_frequency` as a variable and only log every `accumulation_steps`?